### PR TITLE
Made Annoy generic so to work with Vector whos key can be String, Int…

### PR DESCRIPTION
**Pull request by nemo83**
Hi,

first and foremost, thank you for writing this lib. Fantastic job.

I've added `searchK` as parameter of the `Search` as using the default `-1` is suboptimal in some circumstances.

We also use keys different from `Int` so I was writing another layer on top of the lib.. so I thought.. why not implementing it directly in `Annoy4s`?

Feedback welcome.

### CODE REVIEW

File: src/main/scala/annoy4s/Annoy.scala [hunk 1]
```diff
@@ -14,76 +14,103 @@
...
+  indexToId: Seq[T],
   annoyIndex: Pointer,
   dimension: Int
 ) {
   def close() = {
     Annoy.annoyLib.deleteIndex(annoyIndex)
   }
 
-  def query(vector: Seq[Float], maxReturnSize: Int) = {
+  def query(vector: Seq[Float], maxReturnSize: Int, searchK: Int) = {
```
Comment (kevoriordan) [inline src/main/scala/annoy4s/Annoy.scala]: worth setting a default value of searchK: Int = -1 here? That way preserving backwards compatibility and making it easier for people who don't care about searchK to use the api.

File: src/main/scala/annoy4s/Annoy.scala [hunk 2]
```diff
@@ -14,76 +14,103 @@
...
     val result = Array.fill(maxReturnSize)(-1)
     val distances = Array.fill(maxReturnSize)(-1.0f)
-    Annoy.annoyLib.getNnsByVector(annoyIndex, vector.toArray, maxReturnSize, -1, result, distances)
+    Annoy.annoyLib.getNnsByVector(annoyIndex, vector.toArray, maxReturnSize, searchK, result, distances)
     result.toList.filter(_ != -1).map(indexToId.apply).zip(distances.toSeq)
   }
 
-  def query(id: Int, maxReturnSize: Int) = {
+  def query(id: T, maxReturnSize: Int, searchK: Int) = {
```
Comment (kevoriordan) [inline src/main/scala/annoy4s/Annoy.scala]: as above

File: src/main/scala/annoy4s/Annoy.scala (hunk 1 above)
Comment (pvinchon) [inline src/main/scala/annoy4s/Annoy.scala]: Would it make sense to keep the default `-1` here ? Similar to annoy python

File: src/main/scala/annoy4s/Annoy.scala (hunk 2 above)
Comment (pvinchon) [inline src/main/scala/annoy4s/Annoy.scala]: ditto

File: src/main/scala/annoy4s/Annoy.scala (hunk 1 above)
Comment (JasonTam) [inline src/main/scala/annoy4s/Annoy.scala]: consider `searchK=-1` default param

**Review by kevoriordan — APPROVED** [review]
2 small comments, otherwise lgtm

File: src/main/scala/annoy4s/Annoy.scala [hunk 3]
```diff
@@ -14,76 +14,103 @@
...
+import annoy4s.Converters.KeyConverter
 import better.files._
+import com.sun.jna._
+
 import scala.io.Source
 
-class Annoy(
-  idToIndex: Map[Int, Int],
-  indexToId: Seq[Int],
+class Annoy[T](
```
Comment (JasonTam) [inline src/main/scala/annoy4s/Annoy.scala:25]: consider default `Int` type and make the generic in some other class/trait. (for backwards compat)

**Review by pvinchon — APPROVED** [review]
One suggestion, looks good otherwise

**Comment by nemo83** [comment]
Great points every one, but unfortunately cannot implement any of them:
1) setting searchK to -1 will lead to 
> multiple overloaded alternatives of method query define default arguments
2) I cannot default `T` in `Annoy[T]` to `Int`, I should indeed create a new class and it feels like an overkill

**Comment by pishen** [comment]
Try to implement the "default to -1" in https://github.com/pishen/annoy4s/pull/2/commits/642c1e401cf09221e693ad856efd2a07d856fe45, would it be OK like this?

Just changed the test back to see if it work, feel free to modify it or add more test for searchK if you want.
And, thanks for the contribution! This looks cool to me. Will wait for a while to see if there are further comments.

**Comment by nemo83** [comment]
Your changes in making `searchK` to default to `-1` LGTM.

I will try to implement some meaningful test for `searchK` tomorrow.

Thanks for the quick review.

**Comment by nemo83** [comment]
Added a simple test regarding searchK. Happy for the PR to be merged.

**Pull request merged**